### PR TITLE
Prepull container images hosted in docker hub

### PR DESCRIPTION
**Pull request by fmuyassarov**
Prepull container images hosted in docker hub
Fixes: metal3-io/project-infra#166

### CODE REVIEW

File: ci/scripts/image_scripts/provision_metal3_image.sh [hunk 1]
```diff
@@ -59,6 +59,18 @@ if [[ "${DEPLOY_METAL3}" == "true" ]]; then
   rm -rf "${M3_DENV_PATH}"
 fi
 
+export KUBERNETES_VERSION=${KUBERNETES_VERSION:-"v1.20.2"}
+# Container images that we prepul into the CI image.
```
Comment (furkatgofurov7) [inline ci/scripts/image_scripts/provision_metal3_image.sh:63]: `pre-pull`

File: ci/scripts/image_scripts/provision_metal3_image.sh [hunk 2]
```diff
@@ -59,6 +59,18 @@ if [[ "${DEPLOY_METAL3}" == "true" ]]; then
...
+export IMG_REGISTRY="${IMG_REGISTRY:-"docker.io/registry:latest"}"
+export IMG_GOLANG_IMG="${IMG_GOLANG_IMG:-"registry.hub.docker.com/library/golang:1.15.3"}"
+export IMG_CENTOS_IMG="${IMG_CENTOS_IMG:-"docker.io/centos:centos8"}"
+export IMG_KIND_NODE_IMAGE="${IMG_KIND_NODE_IMAGE:-"kindest/node:${KUBERNETES_VERSION}"}"
+
+# Download contaainer images.
```
Comment (furkatgofurov7) [inline ci/scripts/image_scripts/provision_metal3_image.sh:69]: `container`

File: ci/scripts/image_scripts/provision_metal3_image_centos.sh [hunk 3]
```diff
@@ -51,6 +51,17 @@ if [[ "${DEPLOY_METAL3}" == "true" ]]; then
   rm -rf "${M3_DENV_PATH}"
 fi
 
+export KUBERNETES_VERSION=${KUBERNETES_VERSION:-"v1.20.2"}
+# Container images that we prepul into the CI image.
```
Comment (furkatgofurov7) [inline ci/scripts/image_scripts/provision_metal3_image_centos.sh:55]: same here, please see above comment.

File: ci/scripts/image_scripts/provision_metal3_image_centos.sh [hunk 4]
```diff
@@ -51,6 +51,17 @@ if [[ "${DEPLOY_METAL3}" == "true" ]]; then
...
+export IMG_REGISTRY="${IMG_REGISTRY:-"docker.io/registry:latest"}"
+export IMG_GOLANG_IMG="${IMG_GOLANG_IMG:-"registry.hub.docker.com/library/golang:1.15.3"}"
+export IMG_CENTOS_IMG="${IMG_CENTOS_IMG:-"docker.io/centos:centos8"}"
+
+# Download contaainer images.
```
Comment (furkatgofurov7) [inline ci/scripts/image_scripts/provision_metal3_image_centos.sh:60]: same

**Review by furkatgofurov7 — COMMENTED** [review]
What do you think of putting pulling images in a separate script (i.e the same way we have reset_cloud_init.sh) and calling that script from the scripts that you are changing now? That way IMO, we would avoid repetition and changing each script to add new image in the future. 

Small typo nits inline:

**Comment by fmuyassarov** [comment]
> What do you think of putting pulling images in a separate script (i.e the same way we have reset_cloud_init.sh) and calling that script from the scripts that you are changing now? That way IMO, we would avoid repetition and changing each script to add new image in the future.
> 
> Small typo nits inline:

True. I was thinking the same in the beginning, but the problem is that, we have different set of images for centos and ubuntu. So, having one common script will cause to download some images that are not even needed by another OS.

**Comment by fmuyassarov** [comment]
@furkatgofurov7 thanks, fixed the typos.

**Comment by furkatgofurov7** [comment]
> we have different set of images for centos and ubuntu

I can only see kind image being pulled in case of ubuntu, otherwise seems to be all the same in both ubuntu and centos. Also, why are we pulling centos image in both ubuntu&centos?

**Comment by fmuyassarov** [comment]
> > we have different set of images for centos and ubuntu
> 
> I can only see kind image being pulled in case of ubuntu, otherwise seems to be all the same in both ubuntu and centos. 

Yes true. I'm okay with any options. If you are +1 for a separate script I can do it.
> Also, why are we pulling centos image in both ubuntu&centos, btw?

Centos is needed for the cases when we will need to build container while running CI in Ironic repos.

**Review by kashifest — COMMENTED** [review]
why are we fixing versions? why not latest?

**Comment by fmuyassarov** [comment]
> why are we fixing versions? why not latest?

You mean for all images? or only Calico?

**Comment by kashifest** [comment]
> > why are we fixing versions? why not latest?
> 
> You mean for all images? or only Calico?

calico

**Comment by fmuyassarov** [comment]
> > > why are we fixing versions? why not latest?
> > 
> > 
> > You mean for all images? or only Calico?
> 
> calico

Because if you look at the calico [mainfest yaml](https://docs.projectcalico.org/manifests/calico.yaml) which we are using [here](https://github.com/metal3-io/metal3-dev-env/blob/7ab1eff181df5b9db21ad9eab8e63c04b24c7281/vm-setup/roles/v1aX_integration_test/tasks/verify.yml#L44) , it is using images with the latest release tag (v3.17.1). If we download latest, then crictl will ignore the images with latest tag, and try to pull the specific version.

**Comment by kashifest** [comment]
> > > > why are we fixing versions? why not latest?
> > > 
> > > 
> > > You mean for all images? or only Calico?
> > 
> > 
> > calico
> 
> Because if you look at the calico [mainfest yaml](https://docs.projectcalico.org/manifests/calico.yaml) which we are using [here](https://github.com/metal3-io/metal3-dev-env/blob/7ab1eff181df5b9db21ad9eab8e63c04b24c7281/vm-setup/roles/v1aX_integration_test/tasks/verify.yml#L44) , it is using images with the release tag (v3.17.1). If we download latest, then crictl will ignore the images with latest tag, and try to pull the specific version.

Aha ! ok

**Review by kashifest — DISMISSED** [review]

**Comment by furkatgofurov7** [comment]
> Yes true. I'm okay with any options. If you are +1 for a separate script I can do it.

The benefit of separating it, was mainly for keeping the list of images up-to-date in one place(script), rather than going in each and every image building script, when we have to add new image. Otherwise I will leave this up to you/other reviewers to decide.

**Comment by kashifest** [comment]
> > Yes true. I'm okay with any options. If you are +1 for a separate script I can do it.
> 
> The benefit of separating it, was mainly for keeping the list of images up-to-date in one place(script), rather than going in each and every image building script, when we have to add new image. Otherwise I will leave this up to you/other reviewers to decide.

+1 for separate script. Since its cosmetic change, you can also do it in a separate PR later IMO

**Comment by fmuyassarov** [comment]
@kashifest  , @furkatgofurov7 I have a question. I have added centos image downloading both centos and ubuntu. Centos is mainly Ironic repo., However, in Ironic repos we only run ubuntu integration test. So, do you think I should keep centos image  dowload for centos CI image as well ( in case someone triggers the centos job in Ironic repo), or shall I remove it to avoid extra space consumption.

**Comment by kashifest** [comment]
> @kashifest , @furkatgofurov7 I have a question. I have added centos image downloading both centos and ubuntu. Centos is mainly Ironic repo., However, in Ironic repos we only run ubuntu integration test. So, do you think I should keep centos image dowload for centos CI image as well ( in case someone triggers the centos job in Ironic repo), or shall I remove it to avoid extra space consumption.

I am neutral on this, anything is ok for me.

**Comment by fmuyassarov** [comment]
I'm also +1 for a clean script. But if I do it in an external script, where should I add kind image downloading? As part of the provision_metal3_image.sh ?

**Comment by furkatgofurov7** [comment]
> I am neutral on this, anything is ok for me.

+1, as far as I do not have strong opinion on this.

**Comment by furkatgofurov7** [comment]
> I'm also +1 for a clean script. But if I do it in an external script, where should I add kind image downloading? As part of the provision_metal3_image.sh ?

can't we condition that with `IMAGE_OS` in the same separate script, since it will be exported in both [centos](https://github.com/Nordix/airship-dev-tools/blob/master/ci/scripts/image_scripts/provision_metal3_image_centos.sh#L18) and [ubuntu](https://github.com/Nordix/airship-dev-tools/blob/master/ci/scripts/image_scripts/provision_metal3_image.sh#L30) provision scripts?

**Comment by furkatgofurov7** [comment]
> +1 for separate script. Since its cosmetic change, you can also do it in a separate PR later IMO

+1 for me.

**Comment by furkatgofurov7** [comment]
LGTM.

**Comment by fmuyassarov** [comment]
@kashifest @furkatgofurov7 I separated the script. Please take a look once more. Thank you.

**Review by kashifest — APPROVED** [review]

**Review by smoshiur1237 — COMMENTED** [review]
Looks good to me.

**Review by namnx228 — COMMENTED** [review]
LGTM. Just one question:Why do we use the Golang version 1.15.3 instead of the latest one?

**Comment by fmuyassarov** [comment]
> LGTM. Just one question:Why do we use the Golang version 1.15.3 instead of the latest one?

Because this is the version used in CAPM3 and BMO [dockerfiles](https://github.com/metal3-io/cluster-api-provider-metal3/blob/5a09b475a0c00f5349b7a85b181e01c03abe6dfc/Dockerfile#L16).

**Pull request merged**